### DISCUSSION TITLE
How do you send an array of collections as input to a stored procedure?

### DISCUSSION THREAD

**Discussion by AndreFK** [godror/godror#309] [Q&A] [answered]
I have been stuck on this for a week now. I have this stored procedure

```SQL
PROCEDURE submit_rates
(
    pi_user_email   IN    xdb.users.email%TYPE,
    pi_array_ratecodes    IN    rms.RMS_Common_Pkg.recordset_ratecodes,
    po_status   OUT   VARCHAR2,
    po_message    OUT   VARCHAR2
)

/*
PI_ARRAY_RATECODES          IN         This is mandatory IN parameter for rates code columns table 
                                       and the value should be passed as an array for one row only 
                                       to  insert or update in RATECODES table. Pass value  for 4 
                                       columns as follows:
                                       1)  RATECODE_ID         - Integer PK - (Blank for new record)
				       2)  RATECODE            - String max 20 Characters - Mandatory
                                       3)  RATECODE_NAME       - String max 100 Characters
                                       4)  RATE_STRUCTURE      - String max 20 Characters
*/
```
I am not sure to send an array of collections as input to a stored procedure with godror. I already checked another discussion on sending Collections as parameters and tried doing something like this:

```GO

const SaveRatePlan = `begin rms.RMS_Submit_data_Pkg.SUBMIT_RATES(:1, :2, :3, :4); end;`

type RatePlan struct {
	UserEmail         *string            `json:"userEmail"`
	RateCodes         []RateCode         `json:"rateCodes"`
}

type RateCode struct {
	RateCodeId        *int    `json:"rateCodeId" db:"RATECODE_ID"`
	RateCode          *string `json:"rateCode" db:"RATECODE"`
	RateCodeName      *string `json:"rateCodeName" db:"RATECODE_NAME"`
	RateStucture      *string `json:"rateStucture" db:"RATE_STRUCTURE"`
}
        
func (o *Oracle) CreateRatePlan(ctx context.Context, ratePlan entity.RatePlan) (int, error) {

	var status, message string

	params := make([]interface{}, 0)
	params = append(params, godror.PlSQLArrays)
	params = append(params, ratePlan.UserEmail)
	params = append(params, ratePlan.RateCodes)
	params = append(params, sql.Out{Dest: &status})
	params = append(params, sql.Out{Dest: &message})

	_, err := o.db.ExecContext(ctx, SaveRatePlan, params...)

	if err != nil {
		return 0, err
	}

	log.Println("status: ", status)
	log.Println("message: ", message)

	return 1, nil
}
```

This code returns the following error: ```"getObjectType(\"\") conn=0x10260ba10: ORA-22303: OCI-22303: type \"\".\"\" not found"```

Can anyone tell me what I'm doing wrong and point me towards the right direction please?

**Comment by tgulacsi** [comment, marked as answer]
TL;DR;
See TestInputWithNativeSlice in z_plsql_types_test.go, and the new TestArrayOfRecords at fe4f081 .

The answer for your question ("what I'm doing wrong") is that you must inform godror somehow 
(add godror.ObjectTypeName to yout struct) that what is the type's name (and also the fields' db name - see the "godror" struct field tags).
Or you have to build up your DB object "by hand" as in TestInptuWithNativeSlice is done (ot := conn.GetObjectType; ot.NewCollection; ...).
This gets complicated (you have to GetObjectType for each used object type, 
create the object/collection, fill its fields/add items.
All this in  dependency (reverse) order (the innermost first, then its container, etc).

This is what the godror.TypeObjectName embedding saves you from.
